src/mpid/ch4/generic/am/mpidig_am_msg.h

```diff
@@ -77,13 +77,21 @@ MPL_STATIC_INLINE_PREFIX void MPIDIG_recv_copy(void *in_data, MPIR_Request * rre
         /* otherwise if recv size = 0, it is at least a truncation error */
         MPIR_STATUS_SET_COUNT(rreq->status, 0);
     } else if (p->recv_type == MPIDIG_RECV_DATATYPE) {
+        MPI_Aint max_data_size;
+        MPIR_Datatype_get_size_macro(MPIDIG_REQUEST(rreq, datatype), max_data_size);
+        max_data_size *= MPIDIG_REQUEST(rreq, count);
+
         MPI_Aint actual_unpack_bytes;
         MPIR_Typerep_unpack(in_data, in_data_sz,
                             MPIDIG_REQUEST(rreq, buffer),
                             MPIDIG_REQUEST(rreq, count),
                             MPIDIG_REQUEST(rreq, datatype), 0, &actual_unpack_bytes);
-        if (in_data_sz > actual_unpack_bytes) {
+        if (in_data_sz > max_data_size) {
             rreq->status.MPI_ERROR = MPIDIG_ERR_TRUNCATE(actual_unpack_bytes, in_data_sz);
+        } else if (in_data_sz > actual_unpack_bytes) {
+            rreq->status.MPI_ERROR =
+                MPIR_Err_create_code(MPI_SUCCESS, MPIR_ERR_RECOVERABLE, __func__, __LINE__,
+                                     MPI_ERR_TYPE, "**dtypemismatch", 0);
```

Comment:
This is the case when `MPIR_Typerep_unpack` disagree with `MPIR_Datatype_get_size_macro`? Wouldn't that mean some code error in our implementation rather than a user error? In which case, I think an assertion is more appropriate.

Reply:
No, this happens when the user uses a different datatype for send and receive.  We should return errors where possible.  There's a whole upper level logic in the MPI layer that deals with aborting vs. returning errors.

Reply:
I don't quite get it. We are unpacking from a byte stream to receiver datatype, where does send datatype come in?

Reply:
Oh I see. We only unpack integral data elements, but the send side may send in segments crossing the element boundary, that's the issue, right?